### ISSUE TITLE
Test-ModuleManifest doesn't work when RequiredModules are specified

### ISSUE THREAD

**Issue by cholmes1111** [PowerShell/PowerShell#7722]
This is essentially a duplicate of closed issue [3663](https://github.com/PowerShell/PowerShell/issues/3663) which was marked as closed because while it failed in PowerShell 5.1 as the original poster commented, @SteveL-MSFT,  indicated this works in PowerShell 6. I can not get it to work in 6.0.4.

I'm unclear as to why the required modules have to be installed locally before publishing. As part of the publishing process, a check is performed to ensure that the required modules are available on the repo. If the required modules are not available on the repo, the publishing of the module fails. If the required modules are found on the repo, the publishing of the module will succeed. With that workflow, why require them to be installed locally?
 
Steps to reproduce
------------------

```powershell
New-ModuleManifest m.psd1 -RequiredModules sub
Test-ModuleManifest m.psd1 -verbose

```

Expected behavior
-----------------
I'd expect it to return details on the module being tested. For example: 
```powershell
Test-ModuleManifest m.psd1 -Verbose

ModuleType Version    Name                                ExportedCommands
---------- -------    ----                                ----------------
Manifest   0.0.1      m
```

Actual behavior
---------------
I get an error 
```powershell
Test-ModuleManifest m.psd1 -Verbose
Test-ModuleManifest : The specified RequiredModules entry 'sub' in the module manifest 'C:\Inbox\m.psd1' is invalid. Try again after updating this entry with valid values.
At line:1 char:1
+ Test-ModuleManifest m.psd1 -Verbose
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : ObjectNotFound: (m.psd1:String) [Test-ModuleManifest], DirectoryNotFoundException
+ FullyQualifiedErrorId : Modules_InvalidRequiredModulesinModuleManifest,Microsoft.PowerShell.Commands.TestModuleManifestCommand


ModuleType Version    Name                                ExportedCommands
---------- -------    ----                                ----------------
Manifest   0.0.1      m

```

Environment data
----------------
This happens in version 6.0.4 on both Windows and Ubunutu

Windows
```powershell
> $PSVersionTable
Name                           Value
----                           -----
PSVersion                      6.0.4
PSEdition                      Core
GitCommitId                    v6.0.4
OS                             Microsoft Windows 10.0.14393
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

Ubuntu
```powershell
> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      6.0.4
PSEdition                      Core
GitCommitId                    v6.0.4
OS                             Linux 4.4.0-28-generic #47-Ubuntu SMP Fri Jun...
Platform                       Unix
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0

```

Workaround
------------------
To get around this limitation, we read in the data file information, create an array of the required modules and install each of them.

```powershell
    # Get .psd1 data
    $data = Import-PowerShellDataFile .\*.psd1

    # Get the RequiredModules
    $requiredModules = $data.RequiredModules.ModuleName
    
    if ($requiredModules) {
        # Create credentials used to connect to private NuGet feed
        Write-Output 'Creating credentials'
        $pwd = ConvertTo-SecureString $NuGetApiKey -AsPlainText -Force
        $cred = New-Object Management.Automation.PSCredential ("username", $pwd)
        # Install the required modules
        foreach ($module in $requiredModules) {
            Install-Module $module -Repository $repo -Credential $cred -Verbose
        }
    }
```
While this works, it adds additional, unneeded overhead.

**Comment by cholmes1111** [comment]
Another quick follow up, when installing the required modules, you'll notice that I don't specify a version I always just grab the latest. The locally installed version of the required modules does NOT have to match the version specified in the required modules section of the .psd1. 

If the workflow is going to force you to have the required modules installed locally, it should additionally ensure it's the correct version.

**Comment by mklement0** [comment]
@cholmes1111: 

Good point about not enforcing the version; that bug has already been reported: #7495

That same report, created by @rjmholt, also suggests the following enhancement, which addresses your main issue:

> Provide an option to only check the validity of the manifest itself, and not whether it will work in the current session (this last point I think is part of the need to separate "Module manifests that are not well-formed" from "Module manifests that cannot be loaded in the current session"

Note that this means that the default behavior would remain as-is (except for the need to fix the version bug), and that performing a test for mere well-formedness would be _opt-in_.

---

In the meantime, you can use the stopgap demonstrated in [this SO answer](https://stackoverflow.com/a/52212191/45375).

**Comment by ShaunInIdaho** [comment]
If you're like me and you've come across this page while searching for a solution to publishing modules from a container (Bitbucket/Azure DevOps) I have a very simple work around to this issue.  

# Problem
The problem is that Publish-Module runs Test-ModuleManifest which expects any requiredModules (which will include Externally Required modules) to be installed on the local file system.  This is especially challenging when your module depends on modules from multiple external repos.

# Solution
The thing is that Test-ModuleManifest just looks for the existence of the .psd1 file and doesn't care about the contents.  I was able to work around this by simply identifying the location where modules are installed by default and then creating the module directory and an empty .psd1 file like this:

```
Function Build-RequiredModuleFiles {

  # Get .psd1 data
  $Data = Import-PowerShellDataFile .\*.psd1

  # Get the RequiredModules
  [array]$RequiredModules = $data.RequiredModules

  
  If ($RequiredModules) {

    Set-Location '/root/.local/share/powershell/Modules'

    # Create the required module manifests
    ForEach ($module in $requiredModules) {
      $moduleName = $module.ModuleName
      New-Item $moduleName -type Directory
      Write-Output "Creating empty .psd1 file for module $modulename at $((Get-Location).Path)\$moduleName\$moduleName.psd1"

      # Create manifest
      New-Item ".\$moduleName\$moduleName.psd1"
    }
  }
}
```

**Comment by iSazonov** [comment]
I think it makes no sense to create a module manifest if it can not be used or tested locally.

**Comment by nathan-alden-hp** [comment]
@iSazonov I agree, but unfortunately _some_ hack is necessary due to Microsoft's overly-restrictive coding of `Publish-Module` and the underlying inflexibility of `Test-ModuleManifest`. This is actually a great workaround.

**Comment by rjmholt** [comment]
> I think it makes no sense to create a module manifest if it can not be used or tested locally. Don't add RequiredModules if you have to ignore it on next step.

I would actually prefer `Test-ModuleManifest` to allow levels of validation. I think that especially makes sense for development scenarios with staged testing and publication.

Unfortunately the logic for all of this is currently bound in the largest method in the codebase (the only one that Visual Studio gives a code quality score of 0):

https://github.com/PowerShell/PowerShell/blob/a8213b54b3f061e3828b0f32c512814022002b71/src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs#L1461-L3523

I would like for us to refactor that method to enable a number of scenarios, but the module code is sensitive and services a lot of different cases, so that's currently blocked on improving testing of module scenarios. Also it hasn't been prioritised.

**Comment by iSazonov** [comment]
> I would like for us to refactor that method **to enable a number of scenarios**, but the module code is sensitive and services a lot of different cases, so that's currently blocked on improving testing of module scenarios. Also it hasn't been prioritised.

I think today is good time to re-think and re-design the area.
Short thoughts:
1. Allow new keywords in psd1 files for new features (PrivateData is ugly)
2. Duplicate its in PrivateData section for Windows PowerShell if needed.
2. For Windows PowerShell create a simple servicing fix to pass such manifest check

This open ways for future improvements and enhancements.

**Comment by rjmholt** [comment]
> Allow new keywords in psd1 files for new features (PrivateData is ugly)

Yeah I'd like to do this too. We can't change old versions, but it would be nice to be tolerant as we move forward.

> Duplicate its in PrivateData section for Windows PowerShell if needed.

Don't quite understand you here. Do you mean that we should continue to support the PrivateData field for backward compatibility? If so, I agree.

> For Windows PowerShell create a simple servicing fix to pass such manifest check

This one I'm not sure we can do. To be clear I have no control over that, but I think one of the major issues is that it would mark a significant change in behaviour in what's supposed to be a stable product version. Explaining to people that their module does or doesn't load because of a Windows patch would be a nightmare (we already see things like that and it's certainly hard to debug and often harder to explain).

**Comment by iSazonov** [comment]
> Don't quite understand you here. Do you mean that we should continue to support the PrivateData field for backward compatibility? If so, I agree.

Yes, if a module is only for PowerShell Core we have no need to add the section, but if a developer want to make its module compatible with WinPS one should add PrivateData section and WinPS should silently skip new PowerShell Core keywords out of PrivateData section.

> This one I'm not sure we can do. 

It would be minor _compatibility_ fix. I hope MSFT team can approve this.

**Comment by rjmholt** [comment]
> It would be minor compatibility fix.

I think it would represent a fairly significant behaviour change

**Comment by iSazonov** [comment]
> I think it would represent a fairly significant behaviour change

Why? We need only skip new keyword(s) in psd1, no new behavior will be added.

**Comment by rjmholt** [comment]
If you look at the method [here](https://github.com/PowerShell/PowerShell/blob/a8213b54b3f061e3828b0f32c512814022002b71/src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs#L1461-L3523), it's really not written to make such a change simple or low risk.

> no new behavior will be added.

Well (1) allowing the import of previously invalid PSD1s is a new behaviour, and (2) even without that, it's unintentional new behaviours that are the main risk. Making a change to that part of the code today is already a risky proposition, so backporting it and pushing it out as a patch is hard to justify.

**Comment by vercellone** [comment]
I like @ShaunInIdaho 's workaround.  But, in my scenario (where all the required modules are defined in the same repo) I simply set the $env:PSModulePath variable to include the relevant path.
